src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java

```diff
@@ -114,13 +153,54 @@ public void write(StreamSession session, DataOutputStreamPlus out, int version)
         CassandraStreamHeader.serializer.serialize(header, out, version);
         out.flush();
 
-        CassandraStreamWriter writer = header.compressionInfo == null ?
-                                       new CassandraStreamWriter(sstable, header.sections, session) :
-                                       new CompressedCassandraStreamWriter(sstable, header.sections,
-                                                                           header.compressionInfo, session);
+        IStreamWriter writer;
+        if (shouldStreamFullSSTable())
+        {
+            writer = new CassandraBlockStreamWriter(sstable, session, components);
+        }
+        else
+        {
+            writer = (header.compressionInfo == null) ?
+                     new CassandraStreamWriter(sstable, header.sections, session) :
+                     new CompressedCassandraStreamWriter(sstable, header.sections,
+                                                         header.compressionInfo, session);
+        }
         writer.write(out);
     }
 
+    @VisibleForTesting
+    public boolean shouldStreamFullSSTable()
+    {
+        return isFullSSTableTransfersEnabled && isFullyContained;
+    }
+
+    @VisibleForTesting
+    public boolean fullyContainedIn(List<Range<Token>> requestedRanges, SSTableReader sstable)
+    {
+        if (requestedRanges == null)
+            return false;
+        try (KeyIterator iter = new KeyIterator(sstable.descriptor, sstable.metadata()))
+        {
+            while (iter.hasNext())
+            {
+                DecoratedKey key = iter.next();
+                boolean foundFlag = false;
+                for (Range<Token> r : requestedRanges)
```

Comment:
So in the vnodes case is this going to be 256 ranges by default? Should this be a binary search instead of a scan?

Reply:
I refactored it to use `RangeOwnHelper`.